compiler/next/include/chpl/queries/stringify-functions.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This file implements the generic chpl::stringify
+  as well as specializations for some common types.
+
+  stringify can be used to get a string representation of an object
+  for debugging purposes
+ */
+
+#ifndef CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+#define CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "chpl/util/memory.h"
+#include <set>
+#include <sstream>
+
+
+namespace chpl {
+class Context;
+
+enum StringifyKind {
+  DEBUG_SUMMARY,
+  DEBUG_DETAIL,
+  CHPL_SYNTAX
+};
+
+
+template<typename T> struct stringify {
+  std::string operator()(StringifyKind stringKind, const T& stringMe) const = 0;
```

Comment:
Pasting from slack to not lose track:
How much uglier would it be to have the interface be something like `void operator()(std::ostream &s, StringifyKind, const T&)` so that we can pass cout if its just getting logged and an ostringstream if we want to collect it into a string

Reply:
I am not opposed to doing it that way, just want to better understand the reason/benefit over returning the string for the caller to decide what to do with. Is it motivated by the ["tell-don't-ask"](https://martinfowler.com/bliki/TellDontAsk.html) principle?

Reply:
Yeah I'm thinking about allocations & copies. To get a string out, we just need something like:

```c++
template<typename T>
std::string stringifyString(const T& x, StringifyKind kind) {
  std::ostringstream os;
  stringify<T> stringer;
  stringer(os, kind, x);
  return os.str();
}
```

and maybe you'd want to use the name `stringify` for the above function and maybe put the `struct stringify` in a detail namespace.

I'm not familiar with the linked principle. Most of the time I'm in favor of returning values (in the functional programming sense) but here our common case is dumping to stdout/stderr and we'll probably be doing a lot of it.

Slightly unrelated, but @mppf also brought up in Slack the potential issue with recursive types, where we might want to track which things we've output already. Do we also need to be plumbing a context variable through?

Reply:
> Slightly unrelated, but @mppf also brought up in Slack the potential issue with recursive types, where we might want to track which things we've output already. Do we also need to be plumbing a context variable through?

Because `stringify` is a struct, it can have member fields according to what the implementation needs, including things like tracking a set of what is output.

Reply:
updated to take a `std::ostream` and return `void`